### PR TITLE
Support image input for OpenAI chat compatibility

### LINKED ISSUE #3690: Support vision models (image input) in OpenAI API chat completions
### What are you trying to do?

I would like to use the Ollama implemented OpenAI chat completions API and [OpenAI python client](https://github.com/openai/openai-python) to ask question about images (e.g. [llava](https://ollama.com/library/llava) multimodal model).

### How should we solve this?

The official [OpenAI API chat completions](https://platform.openai.com/docs/guides/vision?lang=curl) endpoint (`/v1/chat/completions`) supports sending images with the prompt using `image_url`:

If implemented in Ollama, this curl command should work to send images with prompts:
```sh
curl http://localhost:11434/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "llava",
    "messages": [
      {
        "role": "user",
        "content": [
          {
            "type": "text",
            "text": "What'\''s in this image?"
          },
          {
            "type": "image_url",
            "image_url": {
              "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }'
```

Currently, the ollama server responds with the following error ([thrown in the ollama/openai/openai.go code](https://github.com/ollama/ollama/blob/9df6c85c3a51ce00d6a65be9dd8a06af07b24af5/openai/openai.go#L92)):

```
{"error":{"message":"json: cannot unmarshal array into Go struct field Message.messages.content of type string","type":"invalid_request_error","param":null,"code":null}}(env) 
```

### What is the impact of not solving this?

Implementing this would help to make vision tools built on OpenAI API compatible with Ollama.

A workaround is to use [Ollama Python client](https://github.com/ollama/ollama-python) to send images or use the `/api/generate` endpoint as outlined on the [ollama llava model page](https://ollama.com/library/llava).

Example using Ollama Python client:
```
import ollama
import httpx

response = httpx.get('https://cdn.arstechnica.net/wp-content/uploads/2022/01/GettyImages-90790890-800x533.jpg')
response.raise_for_status()
image = response.content
response = ollama.generate(model='llava', prompt='Describe this photo in detail.', images=[image], keep_alive=0)
print(response["response"])

```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by royjhan**
Supports passing in base64 encoded image into image_url.

E.g.

```
curl http://localhost:11434/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "llava",
    "messages": [
      {
        "role": "user",
        "content": [
          {
            "type": "text",
            "text": "What'\''s in this image?"
          },
          {
            "type": "image_url",
            "image_url": {
               "url": "'$image'"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }' | jq
```

```
{
  "id": "chatcmpl-659",
  "object": "chat.completion",
  "created": 1719016156,
  "model": "llava",
  "system_fingerprint": "fp_ollama",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": " The image shows a cute cartoon of an animal. It appears to be a dog or similar creature, styled with exaggerated features typical in internet memes. The character has big eyes, a round face, and its arms are raised in the air, as if waving or giving a thumbs-up gesture. There's also some motion blur that gives the impression of movement, suggesting the animal might be jumping or dancing. This kind of image is often used in digital communication to convey emotions or add a playful element to text messages. "
      },
      "finish_reason": "stop"
    }
  ],
  "usage": {
    "prompt_tokens": 1,
    "completion_tokens": 112,
    "total_tokens": 113
  }
}
```

Resolves #3690, #5304

### CODE REVIEW

File: openai/openai.go [hunk 1]
```diff
@@ -148,7 +149,33 @@ func toChunk(id string, r api.ChatResponse) ChatCompletionChunk {
...
+					message := api.Message{Role: msg.Role}
+					switch data["type"] {
+					case "text":
+						if text, ok := data["text"].(string); ok {
+							message.Content = text
+						}
+					case "image_url":
+						if urlMap, ok := data["image_url"].(map[string]any); ok {
+							if url, ok := urlMap["url"].(string); ok {
+								if img, err := base64.StdEncoding.DecodeString(url); err == nil {
```
Comment (jmorganca) [inline openai/openai.go:167]: I think we need to remove the `data:image/jpeg;base64,` prefix here:

```
strings.TrimPrefix(url, "data:image/jpeg;base64,")
```

Source: https://platform.openai.com/docs/guides/vision/uploading-base-64-encoded-images

File: openai/openai.go [hunk 2]
```diff
@@ -148,7 +149,33 @@ func toChunk(id string, r api.ChatResponse) ChatCompletionChunk {
...
+									message.Images = append(message.Images, img)
+								}
+							}
+						}
+					}
+					if message.Content != "" || len(message.Images) > 0 {
```
Comment (jmorganca) [inline openai/openai.go:173]: Is there a case where this `if` statement wouldn't be true?

File: openai/openai.go [hunk 3]
```diff
@@ -27,7 +28,7 @@ type ErrorResponse struct {
 
 type Message struct {
 	Role    string `json:"role"`
-	Content string `json:"content"`
+	Content any    `json:"content"`
```
Comment (jmorganca) [inline openai/openai.go:31]: Let's add some tests for this one, especially since we're now accepting `any` for this field

File: openai/openai.go (hunk 1 above)
Comment (royjhan) [reply openai/openai.go:167]: Great catch, thanks

File: openai/openai.go (hunk 2 above)
Comment (royjhan) [reply openai/openai.go:173]: There's a weird edge case where an assistant message can be passed in without content if tool_calls or function_call is specified, but I can still take it out let me know
Screenshot 2024-06-24 at 11 29 14 AM

File: README.md [hunk 4]
```diff
@@ -182,6 +182,12 @@ $ ollama run llama3 "Summarize this file: $(cat README.md)"
  Ollama is a lightweight, extensible framework for building and running language models on the local machine. It provides a simple API for creating, running, and managing models, as well as a library of pre-built models that can be easily used in a variety of applications.
 ```
 
+### Show model information
```
Comment (jmorganca) [inline README.md:185]: I think this might already be in `main`?

File: gpu/assets.go [hunk 5]
```diff
@@ -77,20 +77,27 @@ func cleanupTmpDirs() {
...
-			pid, err := strconv.Atoi(string(raw))
-			if err == nil {
-				if proc, err := os.FindProcess(pid); err == nil && !errors.Is(proc.Signal(syscall.Signal(0)), os.ErrProcessDone) {
-					// Another running ollama, ignore this tmpdir
-					continue
-				}
-			}
-		} else {
-			slog.Debug("failed to open ollama.pid", "path", d, "error", err)
+		if err != nil {
```
Comment (jmorganca) [inline gpu/assets.go:80]: This might need a rebase – I assume it's not part of this PR 😊

File: openai/openai.go (hunk 1 above)
Comment (jmorganca) [reply openai/openai.go:204]: 🙌

File: openai/openai_test.go [hunk 6]
```diff
@@ -82,6 +84,82 @@ func TestMiddleware(t *testing.T) {
...
+			TestPath: "/api/chat",
+			Handler:  Middleware,
+			Endpoint: func(c *gin.Context) {
+				var chatReq api.ChatRequest
+				if err := c.ShouldBindJSON(&chatReq); err != nil {
+					c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
+					return
+				}
+
+				fmt.Print(chatReq)
```
Comment (jmorganca) [inline openai/openai_test.go:100]: remove me?

File: openai/openai_test.go [hunk 7]
```diff
@@ -27,7 +29,7 @@ func TestMiddleware(t *testing.T) {
 
 	testCases := []testCase{
 		{
-			Name:     "chat handler",
+			Name:     "chat handler (default content)",
```
Comment (jmorganca) [inline openai/openai_test.go:32]:
```suggestion
			Name:     "chat handler",
```

can keep the name as the base case here

File: openai/openai_test.go [hunk 8]
```diff
@@ -82,6 +84,82 @@ func TestMiddleware(t *testing.T) {
 				}
 			},
 		},
+		{
+			Name:     "chat handler (image content)",
```
Comment (jmorganca) [inline openai/openai_test.go:88]:
```suggestion
			Name:     "chat handler with image content",
```

**Comment by jmorganca** [comment]
I think this might need a rebase

**Comment by franz101** [comment]
This is amazing!!

File: openai/openai.go [hunk 9]
```diff
@@ -269,10 +271,50 @@ func toModel(r api.ShowResponse, m string) Model {
...
+						if urlMap, ok := data["image_url"].(map[string]any); ok {
+							if url, ok := urlMap["url"].(string); ok {
+								types := []string{"jpeg", "jpg", "png"}
+								for _, t := range types {
+									url = strings.TrimPrefix(url, "data:image/"+t+";base64,")
+								}
+								if img, err := base64.StdEncoding.DecodeString(url); err == nil {
+									message.Images = append(message.Images, img)
+								}
+							} else {
```
Comment (jmorganca) [inline openai/openai.go:301]: This is a big "pyramid" of if statements – is there a way we can flatten it a bit? Generally if else is returning, you can put it before hand and check the opposite and "short circuit" the return

File: openai/openai_test.go [hunk 10]
```diff
@@ -87,6 +88,80 @@ func TestMiddleware(t *testing.T) {
...
+				var chatResp ChatCompletion
+				if err := json.NewDecoder(resp.Body).Decode(&chatResp); err != nil {
+					t.Fatal(err)
+				}
+
+				if chatResp.Object != "chat.completion" {
+					t.Fatalf("expected chat.completion, got %s", chatResp.Object)
+				}
+
+				if chatResp.Choices[0].Message.Content != "Hello! Nice image." {
```
Comment (jmorganca) [inline openai/openai_test.go:160]: Similar to my other feedback. In this expected function let's instead check that the api.ChatRequest is what we want based on the openai input instead of checking .Message.Content since it's a bit confusing

File: openai/openai_test.go [hunk 11]
```diff
@@ -296,3 +371,500 @@ func TestMiddleware(t *testing.T) {
 		})
 	}
 }
+
+const imageEncoding = `iVBORw0KGgoAAAANSUhEUgAAANIAAAB4CAYAAACHHqzKAAAAAXNSR0IArs4c6QAAAIRlWElmTU0AKgAAAAgABQESAAMAAAABAAEAAAEaAAUAAAABAAAASgEb
```
Comment (jmorganca) [inline openai/openai_test.go:375]: Do you have a smaller image? I think a small svg would do

**Comment by royjhan** [comment]
ran cosine similarity with sentence_transformers, resulting in 1.000 for "why is the sky blue?"

File: openai/openai.go [hunk 12]
```diff
@@ -269,10 +271,52 @@ func toModel(r api.ShowResponse, m string) Model {
...
+					return api.ChatRequest{}, fmt.Errorf("invalid message format")
+				}
+				switch data["type"] {
+				case "text":
+					text, ok := data["text"].(string)
+					if !ok {
+						return api.ChatRequest{}, fmt.Errorf("invalid message format")
+					}
+					message.Content = text
+				case "image_url":
```
Comment (jmorganca) [inline openai/openai.go:294]: We can merge this as is, however with @mxyng 's new messages changes that collate messages, it would be better to instead create individual messages for each text and image content type we see

File: openai/openai.go [hunk 13]
```diff
@@ -269,10 +271,52 @@ func toModel(r api.ShowResponse, m string) Model {
...
 	var messages []api.Message
 	for _, msg := range r.Messages {
-		messages = append(messages, api.Message{Role: msg.Role, Content: msg.Content})
+		switch content := msg.Content.(type) {
+		case string:
+			messages = append(messages, api.Message{Role: msg.Role, Content: content})
+		case []any:
+			message := api.Message{Role: msg.Role}
+			for _, c := range content {
+				data, ok := c.(map[string]any)
```
Comment (jmorganca) [inline openai/openai.go:283]: Instead here let's parse out a structured object

```
type ContentPart struct {
    type string
    text string
    imageUrl ImageUrl
}

type ImageUrl struct {
  url string
}
```

Comment (jmorganca) [reply openai/openai.go:283]: Ah nvm, that might be more complex

File: openai/openai.go [hunk 14]
```diff
@@ -269,10 +271,52 @@ func toModel(r api.ShowResponse, m string) Model {
...
+				case "image_url":
+					urlMap, ok := data["image_url"].(map[string]any)
+					if !ok {
+						return api.ChatRequest{}, fmt.Errorf("invalid message format")
+					}
+					url, ok := urlMap["url"].(string)
+					if !ok {
+						return api.ChatRequest{}, fmt.Errorf("invalid message format")
+					}
+					types := []string{"jpeg", "jpg", "png"}
```
Comment (jmorganca) [inline openai/openai.go:303]: to confirm, are these the only types we support?

File: openai/openai_test.go [hunk 15]
```diff
@@ -269,3 +314,5 @@ func TestMiddlewareResponses(t *testing.T) {
 		})
 	}
 }
+
+const imageEncoding = `iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII=`
```
Comment (jmorganca) [inline openai/openai_test.go:318]: This might be better at the top of the page instead of hiding down here

**Review by jmorganca — APPROVED** [review]

**Comment by jmorganca** [comment]
LGTM, let's make sure to try this with some sample code from the OpenAI sdk

File: openai/openai.go (hunk 12 above)
Comment (royjhan) [reply openai/openai.go:294]: will keep in mind for a potential refactor

**Comment by royjhan** [comment]
Screenshot 2024-07-12 at 12 52 11 PM

usage with chatbot ui pointing openai to localhost

File: openai/openai.go [hunk 16]
```diff
@@ -269,10 +271,69 @@ func toModel(r api.ShowResponse, m string) Model {
...
+				}
+				switch data["type"] {
+				case "text":
+					text, ok := data["text"].(string)
+					if !ok {
+						return nil, fmt.Errorf("invalid message format")
+					}
+					message.Content = text
+				case "image_url":
+					urlMap, ok := data["image_url"].(map[string]any)
```
Comment (jmorganca) [inline openai/openai.go:295]: var url string
if urlMap, ok := data["image_url"].(map[string]any); ok {
  url = ..
}

**Comment by pulinagrawal** [comment]
@royjhan Thanks for your work on this. 

I am not sure if this completely resolves #5304 though. The endpoint implemented  here is completions. To truly resolve #5304 the embedding endpoint will also need to updated to support image only and image+text inputs.

**Comment by heshengtao** [comment]
Dear experts, when will the OpenAI image input feature be updated to the Windows application version? Thank you very much for your contributions!

**Comment by franz101** [comment]
Testing ir here with qwen2 I'm running into issues. Is it fully supported by all vision models?
https://colab.research.google.com/drive/1KcYpuuRbNMtoTaQOmOdlZ3h-66VoL0df?usp=sharing

**Comment by royjhan** [comment]
@franz101 image input is supported for these models: https://ollama.com/search?q=&c=vision
